internal/queue/ctqueue.go

```diff
@@ -0,0 +1,120 @@
+package queue
+
+import (
+	"sync"
+	"sync/atomic"
+)
+
+// CTQueue is a thread safe queue implementation which allows safe usage among different threads/go-routines and can
+// be closed at any given time. The queue can be closed by either allowing the remaining elements to remain within or
+// to be extracted. Once the queue is closed, no new elements may be added, but existing elements can still be popped of
+// the queue.
+type CTQueue[T any] struct {
+	items  []T
+	cond   *sync.Cond
+	closed int32
```

Comment:
Go 1.19 beta includes an atomic bool type. Let's keep this as an int32 for now but remember to swap it for an atomic bool when 1.19 goes to stable (or at least release candidate, such that it's available for download on the go homepage).

cf https://pkg.go.dev/sync/atomic@master#Bool